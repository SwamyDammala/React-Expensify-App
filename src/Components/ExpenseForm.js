import React from 'react'
import moment from 'moment'
import { SingleDatePicker} from 'react-dates'


const now=moment();
// console.log(now.format('MMM Do, YYYY'));


class ExpenseForm extends React.Component{
    constructor(props){
        super(props)    
            this.state={
                    description:props.expense?props.expense.description:'',
                    notes:props.expense?props.expense.notes:'',
                    amount:props.expense?(props.expense.amount/100).toString():'',
                    createdAt:props.expense?moment(props.expense.createdAt):moment(),
                    calendarFocused:false,
                    error:''
                }
        }
     
    handleDescription=(e)=>{
        const description=e.target.value
        this.setState(()=>({description}))
    }

    handleChange=(e)=>{
        const amount=e.target.value
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(()=>({amount}))
        }        
    }

    onDateChange=(createdAt)=>{
        if(createdAt){
            this.setState(()=>({createdAt}))
        }
    }

    onFocusChange=({focused})=>{
        this.setState(()=>({calendarFocused:focused}))
    }

    handleNotes=(e)=>{
        const notes=e.target.value
        this.setState(()=>({notes}))
    }

    handleSubmit=(e)=>{
        e.preventDefault();
        if(!this.state.description || !this.state.amount){
            this.setState(()=>({error:'Please provide the description and amount'}))
        }
        else{
            this.setState(()=>({error:''}))
            this.props.onSubmit({
                description:this.state.description,
                amount:parseFloat(this.state.amount,10)*100,
                createdAt:this.state.createdAt.valueOf(),
                notes:this.state.notes
            })
        }
    }

    render(){
        return(
            <div>
                
                <form className='form' onSubmit={this.handleSubmit}>
                    {this.state.error && <p className='form__error'>{this.state.error}</p>}
                    <input type='text' placeholder='Description'
                     className='text-input'
                     autoFocus value={this.state.description} 
                     onChange={this.handleDescription}/>
                    <input type='number' placeholder='Amount' 
                     className='text-input'
                    value={this.state.amount} onChange={this.handleChange}/>
                    <SingleDatePicker 
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}//show the previous month dates...
                    isOutsideRange={(()=>false)} //it will
                    />
                     <textarea placeholder='Add a note for your expense'
                      className='textarea' 
                     value={this.state.notes} onChange={this.handleNotes}></textarea>
                     <div>
                     <button className='button'>Save Expense</button>
                     </div>
                </form>
            </div>
        )
    }
}

export default ExpenseForm




