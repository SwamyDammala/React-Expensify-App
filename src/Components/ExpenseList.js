import React from 'react'
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import SelectExpenses from '../selectors/expenses'

export const ExpenseList=(props)=>(
    <div>
        {
            props.expenses.length===0?<p>No Expenses</p>:
            props.expenses.map(expense=><ExpenseListItem key={expense.id} {...expense} />)
        }
        
    </div>
)

// const ConnectedExpensesList=connect((state)=>{
//     return {
//         name:'swamy'
//     }
// })(ExpanseList)

const MapStateToProps=((state)=>{
    return {
        expenses:SelectExpenses(state.expenses,state.filters)
    }
})

export default connect(MapStateToProps)(ExpenseList)

