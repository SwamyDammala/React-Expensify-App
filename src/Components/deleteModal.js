import React from 'react'
import Modal from 'react-modal'

const DeleteModal=(props)=>(
    <Modal 
            isOpen={props.deleteConfim}
            onRequestClose={props.handleModal}
            contentLabel='Delete Expense'
            className='modal'>
        
            <p className='modal__body'>Do you want to remove this item from expense tracker?</p>
            <button onClick={props.handleModal}>Cancel</button>
            <button onClick={props.remove}>Delete</button>
    </Modal>
)

export default DeleteModal

