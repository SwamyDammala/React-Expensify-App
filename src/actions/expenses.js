import uuid from 'uuid'

export const addexpense=({
    description='',
    notes='',
    amount=0,
    createdAt=0
            }={}
    )=>({
       type:'Add_expense',
       expense:{
           id:uuid(),
           description,
           notes,
           amount,
           createdAt
       }

})

//Below is the remove the particular expense from expenses array based on the id. We are sending id to match and delete the item from expenses list
export const removeexpense=({id}={})=>({
    type:'Remove_expense',
    id
})

//Below is dunction to edit the exisiting expnase details in expenses array based on the expense id. We are sending update values as object and receive them as objects 
export const editexpense=(id,updates)=>({
    type:'Edit_expense',
    id,
    updates
})