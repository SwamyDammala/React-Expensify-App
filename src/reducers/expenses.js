const expenseReducerDefaultState=[]

//Below is the function we will pass into create store for expense array to update based on diferent types of conditions
const expenseReducer=(state=expenseReducerDefaultState,action)=>{
    switch(action.type){
                case 'Add_expense':            
                    return  [
                            ...state,
                            action.expense
                            ]
                case 'Edit_expense':
                        return state.map((expense)=>{
                            if(expense.id===action.id){
                                return{
                                    ...expense,
                                    ...action.updates
                                     } 
                                }
                            else{
                                return expense
                                }
                            })
                case 'Remove_expense':
                        return state.filter(({id})=>(action.id!==id))
                case 'SET_EXPENSES':
                    return  action.expenses
                    
                default:
                    return state            
            }
}

export default expenseReducer