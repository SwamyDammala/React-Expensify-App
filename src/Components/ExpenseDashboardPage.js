import React from 'react'
import ExpenseList from './ExpenseList'
import ExpenseListFilter from './ExpenseListFilter'
import ExpenseSummary from './ExpenseSummary'
import DeleteModal from './deleteModal'

const ExpenseDashboardPage=()=>(
    <div>
        <ExpenseSummary />
        <ExpenseListFilter />
        <ExpenseList />
    </div>    
)

export default ExpenseDashboardPage