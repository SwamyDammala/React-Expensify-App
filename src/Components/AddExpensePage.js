import React from 'react'
import ExpenseForm from './ExpenseForm'
import { connect } from 'react-redux'
import { addexpense } from '../actions/expenses'


export class AddexpensePage extends React.Component{
    onSubmit=(expense)=>{
        this.props.addexpense((expense))
        this.props.history.push('/')
    }
    render(){
        return(
            <div>
            <h3>This is new expense page</h3>
            <ExpenseForm onSubmit={this.onSubmit} />
            </div>
        )
    }
}

//This is firstway to dispatch the add expense method with out using maps to props
// const AddexpensePage=(props)=>(
//     <div>
//         <h3>This is new expense page</h3>
//         <ExpenseForm onSubmit={(expense)=>{
//             props.dispatch(addexpense(expense))
//             props.history.push('/')
//         }} />
//     </div>    
// )

const mapsDispatchToProps=(dispatch)=>({
    addexpense:(expense)=>dispatch(addexpense(expense))
})

export default connect(undefined,mapsDispatchToProps)(AddexpensePage)