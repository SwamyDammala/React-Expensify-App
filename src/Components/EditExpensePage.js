import React from 'react'
import ExpenseForm from './ExpenseForm'
import { editexpense } from '../actions/expenses'
import { connect } from 'react-redux'
import { removeexpense } from '../actions/expenses'
import expenses from '../tests/fixtures/expenses'

export class EditexpensePage extends React.Component{
    onSubmit=(expense)=>{
        this.props.editexpense(this.props.expense.id,expense)
        this.props.history.push('/')
    }
    onRemove=()=>{
        this.props.removeexpense({id:this.props.expense.id})
        this.props.history.push('/')
    }
    render(){
        return(
            <div>
            <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit}/>
                <button onClick={this.onRemove}>Remove Expense</button>
                </div>
        )
    }
}

// const EditexpensePage=(props)=>{
//     return(
//         <div>            
//             <ExpenseForm expense={props.expenses}
//                     onSubmit={(expense)=>{
//                         props.dispatch(editexpense(props.expenses.id,expense))
//                         console.log('updated',expense)
//                         props.history.push('/')
//             }}/>
//                 <button onClick={()=>{props.dispatch(removeexpense({id:props.expenses.id}))
//                     props.history.push('/')
//                     }}>Remove Expense</button>
//      </div>
//     ) 
// }

const MapsStoreToProps=(store,props)=>{
    return{
        expense:store.expenses.find((expense)=>(expense.id===props.match.params.id))
    }
}

const mapsDispatchToProps=(dispatch,props)=>{
    return{ 
        editexpense:(id,expense)=>dispatch(editexpense(id,expense)),
        removeexpense:(id)=>dispatch(removeexpense(id))
    }

}

export default connect(MapsStoreToProps,mapsDispatchToProps)(EditexpensePage)
