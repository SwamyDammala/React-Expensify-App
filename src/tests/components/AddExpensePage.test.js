import React from 'react'
import {shallow} from 'enzyme'
import expenses from '../fixtures/expenses'
import {AddexpensePage} from '../../Components/AddExpensePage'


let addexpense,history,wrapper

beforeEach(()=>{
    addexpense=jest.fn()
    history={'push':jest.fn()}
    wrapper=shallow(<AddexpensePage addexpense={addexpense} history={history} />)
})

test('Should Check rendering of  Add Expense in AddExpensePage',()=>{
    expect(wrapper).toMatchSnapshot()
})

test('Should Check handle method of  Add Expense in AddExpensePage',()=>{

    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1])
    expect(history.push).toHaveBeenLastCalledWith('/')
    expect(addexpense).toHaveBeenLastCalledWith(expenses[1]);
})
