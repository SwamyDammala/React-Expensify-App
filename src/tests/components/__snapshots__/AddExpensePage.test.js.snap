// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Check rendering of  Add Expense in AddExpensePage 1`] = `
<div>
  <h3>
    This is new expense page
  </h3>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
